security/pkg/k8s/controller/workloadsecret.go

```diff
@@ -183,10 +196,42 @@ func GetSecretName(saName string) string {
 	return secretNamePrefix + saName
 }
 
+// Determine if the object is "enabled" for Istio.
+// Currently this looks at the object's namespace annotation, but could be
+// extended to allow hierarchy of global < namespace < object annotations/labels.
+func (sc *SecretController) istioEnabledObject(obj metav1.Object) bool {
+	const label = "istio-managed"
+	enabled := !sc.explicitOptIn // for backward compatibility, Citadel always creates secrets
+	// @todo this should be changed to false once we communicate behavior change and ensure customers
+	// correctly mark their namespaces. Currently controlled via command line
+
+	ns, err := sc.core.Namespaces().Get(obj.GetNamespace(), metav1.GetOptions{})
+	if err != nil || ns == nil { // @todo handle errors? Unit tests mocks don't create NS, only secrets
+		return enabled
+	}
+
+	if ns.Labels != nil {
+		if v, ok := ns.Labels[label]; ok {
+			switch strings.ToLower(v) {
+			case "enabled", "enable", "true", "yes", "y":
+				enabled = true
+			case "disabled", "disable", "false", "no", "n":
+				enabled = false
+			default: // leave default unchanged
+				break
+			}
+		}
+	}
+	return enabled
+}
+
 // Handles the event where a service account is added.
 func (sc *SecretController) saAdded(obj interface{}) {
 	acct := obj.(*v1.ServiceAccount)
-	sc.upsertSecret(acct.GetName(), acct.GetNamespace())
+	if _, listed := sc.namespaces[acct.GetNamespace()]; listed ||
```

Comment:
The listwatch solution is different:
https://github.com/istio/istio/pull/11667/files#diff-28afeef34b1bb86d2aa50306457a761e
I think you want to converge to the listwatch solution when merging 1.1 to master? Then this part will be throw-away work.
Also you need to update this line in this PR if you want to use the current approach:
https://github.com/istio/istio/blob/master/security/pkg/k8s/controller/workloadsecret.go#L263

I think it's better to merge https://github.com/istio/istio/pull/11667 into your PR and reuse listwatch. WDYT?

Reply:
Tried merging #11667:
```
$ git fetch upstream pull/11667/head:pr-11667
$ git merge --no-commit pr-11667 | grep -i conflict
CONFLICT (content): Merge conflict in vendor/istio.io/api/python/istio_api/mesh/v1alpha1/proxy_pb2.py
CONFLICT (content): Merge conflict in vendor/istio.io/api/mesh/v1alpha1/proxy.pb.go
CONFLICT (content): Merge conflict in security/pkg/k8s/controller/workloadsecret_test.go
CONFLICT (content): Merge conflict in security/pkg/k8s/controller/workloadsecret.go
CONFLICT (content): Merge conflict in security/cmd/istio_ca/main.go
CONFLICT (content): Merge conflict in pilot/pkg/serviceregistry/kube/controller.go
CONFLICT (content): Merge conflict in install/kubernetes/helm/istio/values.yaml
CONFLICT (content): Merge conflict in install/kubernetes/helm/istio/charts/grafana/dashboards/istio-performance-dashboard.json
CONFLICT (content): Merge conflict in Makefile
CONFLICT (content): Merge conflict in Gopkg.lock
Automatic merge failed; fix conflicts and then commit the result.
```
I'm concerned that the divergence might be large. Perhaps it better to reconsider this work after merging r1.1 to master, at least having a stable baseline would be useful than trying to work with a system that's half changed in it's protobuffers.

Reply:
> The listwatch solution is different

Not sure I understand why the listwatch would require a fundamental change in this PR.
My understanding is the listwatch is an abstraction that aggregates multiple watches internally, one per namespace. 
The listwatch is configured with an array of namespaces to watch. I propose to take that list and use it to populate the `controller.namespaces` map, which is currently populated with one entry only.

Reply:
> Also you need to update this line in this PR if you want to use the current approach:
https://github.com/istio/istio/blob/master/security/pkg/k8s/controller/workloadsecret.go#L263

Correct, need to account for this on recreating a deleted secret. 
BTW - why do we create a secret even when there is no istio annotation on it?
```
saName := scrt.Annotations[ServiceAccountNameAnnotationKey]
if _, error := sc.core.ServiceAccounts(scrt.GetNamespace()).Get(saName, metav1.GetOptions{}); error == nil {
...
```
Is the expectation that getting the SA with "" as the saName fails with an error? Might be better to check if `saName != ""` and bailing out, no?

Reply:
@myidpt can you PTAL?